ocaml-lsp-server/src/dune.ml

```diff
@@ -263,15 +263,11 @@ end = struct
                   match p with
                   | Failed | Interrupted | Success ->
                     let* () =
-                      Document_store.change_all document_store ~f:(fun doc ->
+                      Document_store.parallel_iter document_store ~f:(fun doc ->
                           match Document.kind doc with
-                          | `Other -> Fiber.return doc
+                          | `Other -> Fiber.return ()
                           | `Merlin merlin ->
-                            let doc = Document.update_text doc [] in
-                            let+ () =
-                              Diagnostics.merlin_diagnostics diagnostics merlin
-                            in
-                            doc)
+                            Diagnostics.merlin_diagnostics diagnostics merlin)
                     in
                     Diagnostics.send diagnostics `All
```

Comment:
Does this re-make pipelines and send diagnostics to all files on a dune build completion (when dune is in watch mode)? Sounds super expensive

Reply:
Wait until you see how expensive is the build itself :)

But more seriously, of course it would be better to recompute only the documents which were invalidated by the build. Unfortunately that would be far more difficult.

Reply:
> the documents which were invalidated by the build. Unfortunately that would be far more difficult.

But dune knows what those documents are, right? (because of the incremental re-building it has) 

If yes, that becomes a matter of extending the dune rpc?

Reply:
> the documents which were invalidated by the build. Unfortunately that would be far more difficult.

actually, yes, I started noticing how my laptop fans go off often if I have dune running in watch mode

Reply:
Yes, that would be enough. In particular, the rule would be that a merlin document should be invalidated whenever:

* Any .cmi file in its `-I` directories has been invalidated
* Any of the preprocessing binaries has been invalidated
* Any of the runtime dependencies of the preprocessing binaries has been invalidated

Reply:
> actually, yes, I started noticing how my laptop fans go off often if I have dune running in watch mode

Thankfully you can limit the concurrency with `-j`. That calms down the fans considerably.